File: Depth_First_Search/depth_first_search.py
from .._Graph.graph import *
from collections import deque

def breadth_first_search(G: Graph, src: Vertex) -> list[Vertex]:
    visited: set[Vertex] = set()
    stack: deque[Vertex] = deque()
    result: list[Vertex] = []

    stack.append(src)
    visited.add(src)
    while stack:
        vertex: Vertex = stack.pop()
        result.append(vertex)
        for edge in G.get_adjacency_list()[vertex]:
            neighbor: Vertex = edge.v()
            if neighbor not in visited:
                stack.append(neighbor)
                visited.add(neighbor)
    return result

File: Depth_First_Search/depthFirstSearch.cpp
#include "../_Graph/graph.cpp"
#include <bits/stdc++.h>

using namespace std;

vector<Vertex*> depthFirstSearch(Graph* G, Vertex* src) {
    set<Vertex*> visited;
    stack<Vertex*> s;
    vector<Vertex*> result;

    s.push(src);
    visited.insert(src);
    while (!s.empty()) {
        Vertex* vertex = s.top(); s.pop();
        result.push_back(vertex);
        for (Edge* edge : G->getAdjacencyList()[vertex]) {
            Vertex* neighbor = edge->v;
            if (visited.find(neighbor) == visited.end()) {
                s.push(neighbor);
                visited.insert(neighbor);
            }
        }
    }
    return result;
}
